---
title: "onco"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = F,message = F)
```

#### 0.背景知识一点点

oncoPredict是根据基因表达量来预测药物敏感性的R包。也就是说它可以根据你的样本基因表达量来告诉你每个药物的IC50值，这个值越低就说明药物越管用。

提到药物预测，还有一个pRRophetic包，建议不用看了，因为oncoPredict是它的plus版本。

还有一个cellMiner包，之前写过，可以翻翻看。

#### 1.载入数据

代码参考自：https://mp.weixin.qq.com/s/QRaTd-fIsqq6sPsLmOPvIw，一些背景知识也可以补充下.

在Training Data文件夹下存放着R包作者准备好的数据，用作药物预测的训练集。下载自：https://osf.io/c6tfx/

```{r}
rm(list = ls())
library(oncoPredict)
library(data.table)
library(gtools)
library(reshape2)
library(ggpubr)
dir='./DataFiles/DataFiles/Training Data/'
dir(dir)
```

可以看到其中包括了Cancer Therapeutics Response Portal (CTRP)和Genomics of Drug Sensitivity in Cancer (GDSC)

两个数据库的数据，都是提供了基因表达矩阵和药物IC50表格。


```{r}
exp = readRDS(file=file.path(dir,'GDSC2_Expr (RMA Normalized and Log Transformed).rds'))
exp[1:4,1:4]
dim(exp)
drug = readRDS(file = file.path(dir,"GDSC2_Res.rds"))
drug <- exp(drug) #下载到的数据是被log转换过的，用这句代码逆转回去
drug[1:4,1:4]
dim(drug)
identical(rownames(drug),colnames(exp))
```

drug是药物IC50值，exp是对应细胞系基因的表达矩阵。可以看到二者的样本名称是对应的。

#### 2.操练一下

```{r}
load("../4_tcga_model/TCGA-STAD_sur_model.Rdata")
test = exprSet
```

运行时间很长，所以if(F)注释掉。

```{r}
if(F){
  calcPhenotype(trainingExprData = exp,
                trainingPtype = drug,
                testExprData = test,
                batchCorrect = 'standardize',  #   "eb" for array,standardize  for rnaseq
                powerTransformPhenotype = TRUE,
                removeLowVaryingGenes = 0.2,
                minNumSamples = 10, 
                printOutput = TRUE, 
                removeLowVaringGenesFrom = 'rawData' )
}
```

R包Vignette里关于batchCorrect参数的说明

>batchCorrect options: "eb" for ComBat, "qn" for quantiles normalization, "standardize", or "none"
>
>"eb" is good to use when you use microarray training data to build models on microarray testing data.

>"standardize is good to use when you use microarray training data to build models on RNA-seq testing data (this is what Paul used in the 2017 IDWAS paper that used GDSC microarray to impute in TCGA RNA-Seq data, see methods section of that paper for rationale)

R包Vignette里关于removeLowVaringGenesFrom参数的说明

>#Determine method to remove low varying genes.
#Options are 'homogenizeData' and 'rawData'
#homogenizeData is likely better if there is ComBat batch correction, raw data was used in the 2017 IDWAS paper that used GDSC microarray to impute in TCGA RNA-Seq data.

也就是说，芯片数据就用上面代码里的参数，转录组数据的话，就将batchCorrect改为standardize

removeLowVaringGenesFrom,作者说的也模糊啊。随便吧。

#### 3.看看结果

这是运行之后的结果，被存在固定文件夹calcPhenotype_Output下。文件名也是固定的DrugPredictions.csv。因此一个工作目录只能计算一个数据，你可别混着用哦。

```{r}
testPtype <- read.csv('./calcPhenotype_Output/DrugPredictions.csv', row.names = 1,check.names = F)
testPtype[1:4, 1:4]
dim(testPtype)
identical(colnames(testPtype),colnames(drug))
```

198种药物IC50的预测结果就在这个表格里啦。

#### 结合高低风险画图

```{r}
load("../4_tcga_model/TCGA-STADcox_fp.Rdata")
Group = dat$Risk
Group = factor(Group,levels = c("low_risk","high_risk"))
identical(rownames(testPtype),rownames(dat))
a = apply(testPtype, 2, function(x){
  #x = testPtype[,1]
   wilcox.test(x~Group)$p.value
})
head(a)
#p值最小的10个药物
dg = names(head(sort(a),10))
library(tinyarray)
library(ggplot2)
draw_boxplot(t(testPtype[,dg]),Group)+
  facet_wrap(~ rows,scales = "free")
```

